controllers/subscriber/subscriberControllers.js

```diff
@@ -9,11 +9,39 @@ class BaseController {
 		this._name = name;
 		this._options = options;
 		this._eventName = `d2l-subscribe-${this._name}`;
-		this._subscriptionComplete = Promise.resolve();
 	}
 }
 
 class BaseSubscriber extends BaseController {
+	constructor() {
+		super(...arguments);
+
+		this._subscriptionComplete = Promise.resolve();
+		this._timeouts = new Set();
+	}
+
+	hostDisconnected() {
+		this._timeouts.forEach(timeoutId => clearTimeout(timeoutId));
+	}
+
+	_keepTrying(callback, interval, maxTime, errorPayload, elapsedTime = 0) {
+		const response = callback();
+
+		if (response) return response;
+		if (elapsedTime >= maxTime) {
+			if (this._options.onError) this._options.onError(errorPayload);
+			return response;
+		}
+
+		return new Promise(resolve => {
+			const timeoutId = setTimeout(async() => {
+				this._timeouts.delete(timeoutId);
+				resolve(await this._keepTrying(callback, interval, maxTime, errorPayload, elapsedTime + interval));
```

Comment:
It probably doesn't matter for the purpose of what you're doing here, but "elapsed time" probably doesn't just equal the sum of all the intervals, since the browser will do its best to call your timeout after interval, but there's no guarantee. If accuracy is needed (probably not), then the [user timing API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) may be more appropriate.

Reply:
Thanks, yeah I think this is sufficient for here.